### ISSUE TITLE
Implement SimHash

### ISSUE THREAD

**Issue by XYlearn** [radare/radare2#12530]
**Is your feature request related to a problem? Please describe.**
Afaik, r2 toolchains has no fuzzy comparison method now. SimHash is a good alternative to do fuzzy comparison.

**Describe the solution you'd like**
Implement simhash and make it available in rahash2 radiff2 and radare2.
In rahash2 use `rahash2 -a simhash ...` to calculate simhash of specified content
In radiff2 use `radiff2 -f file1 file2` to compute the difference with simhash algorithm.
In radare2 use `ph simhash` to calculate simhash of a block. 
It's better to make function hash available in radare2.

**Additional context**
Here are some references:
[wiki](https://en.wikipedia.org/wiki/SimHash)
[paper](http://www.cs.princeton.edu/courses/archive/spring04/cos598B/bib/CharikarEstim.pdf)
[application](https://googleprojectzero.blogspot.com/2018/12/searching-statically-linked-vulnerable.html) and it's [repo](https://github.com/googleprojectzero/functionsimsearch)

**Comment by Maijin** [comment]
We have ssdeep in r2extras and r2pm, simhash should also go there.

**Comment by XYlearn** [comment]
> We have ssdeep in r2extras and r2pm, simhash should also go there.

ssdeep is in r2extras because it lives on ssdeep project. But if SimHash is implemented without other dependencies, I think, it should be integrated to r2 toolchains.

**Comment by XVilka** [comment]
@XYlearn - if it will be implemented in C, clean and lean enough - why not? Now the easiest part of the issue - someone should write this code.

**Comment by radare** [comment]
i think its interesting to have context sensitive hash support in r2 core

**Comment by Maijin** [comment]
As long as it's tested and maintained it's ok in core too 👍

**Comment by r4dix10** [comment]
Should we integrate https://github.com/BartMassey/simhash into radare ?

**Comment by radare** [comment]
yes, but without duplicating functionalities, and integrating it with the rest of features in r2, so its not just a copypasta

**Comment by XVilka** [comment]
@r4dix10 note that algorithms are totally different in these - the SimHash from Thomas Dullien consists of these:

 1.   An efficient implementation of a hash function (based on SimHashing) which calculates a 128-bit hash from disassembled functions - and which preserves similarity (e.g. “distance” of two functions can be calculated by simply calculating the hamming distance between two hashes - which translates to two XOR and two POPCNT instructions on x64).
 2.   An efficient search index to allow approximate nearest neighbor search for 128-bit hashes.
 3.   Some supervised machine learning code that can be used to “learn” a good hash function from human-provided examples - given a set of functions that should be similar according to the hash, and a set of functions that should be dissimilar, the code learns a hash function that attempts to respect these examples.

The link you posted is much more basic and simple.

**Comment by r4dix10** [comment]
@XVilka this is what i was thinking to implement:

    pick a hashsize, lets say 32 bits
    let V = [0] * 32 # (ie 32 zeros)
    break the phrase up into features

    irb(main):003:0> 'the cat sat on the mat'.shingles
    => #<Set: {"th", "he", "e ", " c", "ca", "at", "t ",
        " s", "sa", " o", "on", "n ", " t", " m", "ma"}>

    hash each feature using a normal 32-bit hash algorithm

    "th".hash = -502157718
    "he".hash = -369049682
    ...

    for each hash
    if biti of hash is set then add 1 to V[i]
    if biti of hash is not set then take 1 from V[i]
    simhash biti is 1 if V[i] > 0 and 0 otherwise

Now with the generated simhash of two or more files; calculate hamming distance(s).
We may generate feature hash with 128 bit hash algorithm and increase feature length for better performance.

Reference : http://matpalm.com/resemblance/simhash/

**Comment by trufae** [comment]
I have implemented ssdeep algorithm. Is simhash needed too? https://github.com/radareorg/radare2/pull/18490
